File: lists/LinkedList.py
'''
Question

        Implement a singly linked list.

Solution:

        The implementation is based on a node object which is used for the
        entries of the linked list. A head attribute in the List wrapper
        class allows to access the beginning of the list.
'''
class Node(object):

        def __init__(self, data):
                self.data = data
                self.next = None

class List(object):

        def __init__(self):
                self.head = None

        def insert(self, data):

                '''Time Complexity O(1)'''

                node = Node(data)
                if self.head is None:
                        self.head = node
                else:
                        node.next = self.head
                        self.head = node

        def delete(self, data):

                '''Time Complexity O(n)'''

                current = self.head
                prev = None
                while current is not None:
                        if current.data == data:
                                if current == self.head:
                                        self.head = current.next
                                else:
                                        prev.next = current.next
                                break
                        else:
                                prev = current
                                current = current.next

        def toList(self):

                result = []
                current = self.head
                while current is not None:
                        result.append(current.data)
                        current = current.next

                return result

        def __str__(self):

                result = ''
                current = self.head
                while current is not None:
                        result += str(current.data) + ' '
                        current = current.next

                return result

File: lists/linked_list.c
/*
 * Question 1
 *
 *      Implement the data structure singly linked list with the operations
 *      add, delete, find and length in C.
 *
 * Solution
 *
 *      The list is implemented with a struct node. The next pointer of the
 *      last node will be NULL, empty lists are represented by a NULL head
 *      pointer and all nodes will be allocated in the heap.
 *      
 */
#include <assert.h>
#include "linked_list.h"

void add(node* head, int data)
{
    node* current;
    node* node = malloc(sizeof(node));
    node->data = data;

    if (head == NULL)
    {
        head = node;
    }
    else
    {
        current = head;
        while (current->next != NULL)
        {
            current = current->next;
        }
        current->next = node;
    }
}

node* delete(node* head, int data)
{
    node* current = head;
    node* previous = current;
    while (current != NULL)
    {
        if (current->data == data)
        {
            previous->next = current->next;
            if (current == head)
            {
                head = current->next;
            }
            else
            {
                free(current);
            }
            return head;
        }
        previous = current;
        current = current->next;
    }
    return head;
}

/**
 * Question 2
 *
 *      Improve the delete function by implementing it without using a previous
 *      variable to store the last node.
 *
 * Solution
 *
 *      The key to the solution is to use pointers to pointers instead. The
 *      function signature is changed to take a node** as head parameter. This
 *      implementation also yields the advantage, that it does not need to
 *      return the head.
 */
void delete2(node** head, int data)
{
    node** current = head;
    while (*current != NULL)
    {
        node *entry = *current;
        if (entry->data == data)
        {
            *current = entry->next;
            free(entry);
            break;
        }
        else
        {
            current = &entry->next;
        }
    }
}

node* find(node* head, int data)
{
    node* current = head;
    if (current != NULL)
    {
        while (current != NULL)
        {
            if (current->data == data)
            {
                return current;
            }
            current = current->next;
        }
    }
    return NULL;
}

void print_list(node* head)
{
    node* current = head;
    if (current != NULL)
    {        
        while (current->next != NULL)
        {
            printf("%d, ", current->data);
            current = current->next;
        }
        printf("%d\n", current->data);
    }
}
